File: binary-tree-level-order-traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/description/
# BFS
# Time: O(n), Space: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        rList = []
        if root:
            que = collections.deque()
            que.append(root)
            while que:
                curL = []
                qLen = len(que)
                for i in range(qLen):
                    curN = que.popleft()
                    curL.append(curN.val)
                    if curN.left!=None: que.append(curN.left)
                    if curN.right!=None: que.append(curN.right)
                rList.append(curL)                   
        return rList

File: binary-tree-level-order-traversal.java
// https://leetcode.com/problems/binary-tree-level-order-traversal/description/
// BFS
// Time: O(n), Space: O(n)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> rList = new ArrayList<List<Integer>>();
        if(root!=null){
            ArrayDeque<TreeNode> que = new ArrayDeque<TreeNode>();
            que.offer(root);
            while(!que.isEmpty()){
                List<Integer> curL = new ArrayList<Integer>();
                int size = que.size();
                for(int i=0; i<size; i++){
                    TreeNode curN = que.poll();
                    curL.add(curN.val);
                    if(curN.left!=null) que.offer(curN.left);
                    if(curN.right!=null) que.offer(curN.right);
                }
                rList.add(curL);
            }
        }
        return rList;
    }
}
